### ISSUE TITLE
Is there any interest in CFG for native assembly?

### ISSUE THREAD

**Issue by ARF1** [numba/numba#5063]
Diagnosing vectorization failures I found it necessary to look extensively at native assembly and was missing an `inspect_cfg()` for native assembly.

I have [code available](https://gist.github.com/ARF1/0d11db207085f1f1f970dd31857e9a90) that implements this by parsing the assembly. Is there any interest in me putting this in a form suitable for a PR?

For example, for the following `max()` implementation, with this it easy to see that the code is not successfully vectorized even after the native optimization passes:

```python
import numba as nb
import numpy as np

nb.dispatcher._DispatcherBase.inspect_asm_cfg = DummyDispatcher.inspect_asm_cfg

@nb.njit(fastmath=True)
def test_max(arr):
    a = arr.ravel()
    max_value = a[0]
    for i in range(1, a.size):
        max_value = max(max_value, a[i])
    return max_value

test_max(np.random.rand(10))
test_max.inspect_asm_cfg(test_max.signatures[0]).display()
```

Results in:
![cfg](https://user-images.githubusercontent.com/5834577/72170797-d00e4900-33d1-11ea-985f-7afc7e61e0e5.jpg)

**Comment by sklam** [comment]
This is interesting.  Being able to see the loops via the backedges will certainly help.  One concern I have would be if the code is too ad-hoc; e.g. manual parsing for just x86, handling of indirect jumps. I wonder if there are existing tools that we can rely on.  For that, I immediately think of godbolt and they have this issue: https://github.com/mattgodbolt/compiler-explorer/issues/1158. It also points to a llvm tool called llvm-mca (https://llvm.org/docs/CommandGuide/llvm-mca.html), which could be useful.

**Comment by ARF1** [comment]
@sklam Exactly I needed a quick way to identify loops. Only the code in the loops is of interest to me. They are no-trivial to identify in assembly but obvious in CFG.

I am really not very invested in my particular solution to this problem. I cobbled this together because it was quicker than searching for alternatives.

I do however think that CFG for native assembly is really important.

If there are tools to do this more robustly that would certainly be preferable.

---
**Quick summary of the algorithm discussed above**

I agree that the code is likely to be somewhat brittle. But on the other hand since all assembly comes formatted by llvm, parsing it should be a manageable problem in principle. The formatting of the input will be consistent.

Basically the code makes 5 assumptions about the assembly language provided as input that should be reasonably general:

- unconditionals jump are `jmp`
- conditional jumps are `j....`
- function returns are `ret....`
- labels are `[nonwhitespace]:`
- assembler directives begin with a dot and have no colon in their line

**I had not thought about indirect jumps and have no idea how to handle them!**

Luckily I never encountered them in my numba assemblies so far. - That is probably not a good policy for production code though...

**Comment by stuartarchibald** [comment]
I'm very interested in this, and am already heavily invested in analysis tooling e.g. as a public example https://github.com/numba/numba/issues/5028.

@gmarkall suggested [`r2`](https://github.com/radareorg/radare2/) might help, a few minutes with this tool produced this (with a couple of minor edits):

![max_cfg](https://user-images.githubusercontent.com/907717/72433508-5d1a1f00-3791-11ea-9e74-cf1e55b1e31f.png)

**Comment by esc** [comment]
@stuartarchibald can you share  how you managed to create the CFG?

**Comment by stuartarchibald** [comment]
Implemented in https://github.com/numba/numba/pull/5212

**Comment by stuartarchibald** [comment]
From my point of view, this is done.

**Comment by esc** [comment]
@stuartarchibald thanks for following up, @ARF1 do you concur?

**Comment by ARF1** [comment]
@esc Sorry I lost track of this. I am closing this.

**Comment by esc** [comment]
@ARF1 no problem and thanks for following up.
